### PR TITLE
Rewrite sets for statutory sick pay

### PR DESCRIPTION
**Pull request by JonathanHallam**
This replaces the previous unit/integration flow tests with the node
based approach proposed in [ADR 3].

I've had to do something slightly odd with the final outcome nodes because they required some non unique setup steps. Otherwise this is a fairly linear smart answer

Trello - https://trello.com/c/C3MJYR8V/539-smart-answer-flow-tests-calculate-statutory-sick-pay

[ADR 3]: https://github.com/alphagov/smart-answers/blob/main/docs/arch/003-testing-flows.md

⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

### CODE REVIEW

File: test/flows/calculate_statutory_sick_pay_test.rb [hunk 1]
```diff
@@ -0,0 +1,598 @@
...
+
+      should "not allow a sick day in the future" do
+        assert_invalid_response "3000-01-01"
+      end
+
+      should "not allow the linked sickness to end more than 8 weeks before first sick day" do
+        assert_invalid_response "2019-10-01"
+      end
+
+      should "make sure the linked sickness is before first sick day" do
```
Comment (leenagupte) [inline test/flows/calculate_statutory_sick_pay_test.rb:262]:
```suggestion
      should "be invalid if the linked sickness end date is before first sick day" do
```

File: test/flows/calculate_statutory_sick_pay_test.rb [hunk 2]
```diff
@@ -0,0 +1,598 @@
...
+      testing_node :how_often_pay_employee_pay_patterns?
+      add_responses is_your_employee_getting?: "statutory_adoption_pay",
+                    coronavirus_related?: "no",
+                    employee_tell_within_limit?: "yes",
+                    employee_work_different_days?: "no",
+                    first_sick_day?: "2020-01-01",
+                    last_sick_day?: "2020-01-05",
+                    has_linked_sickness?: "yes",
+                    linked_sickness_start_date?: "2019-12-01",
+                    linked_sickness_end_date?: "2019-12-30"
```
Comment (leenagupte) [inline test/flows/calculate_statutory_sick_pay_test.rb:312]: The "assert question rendered" test is missing. When it's added, a response for `paid_at_least_8_weeks?` will need to be added here.

File: test/flows/calculate_statutory_sick_pay_test.rb [hunk 3]
```diff
@@ -0,0 +1,598 @@
...
+    end
+
+    context "next_node" do
+      should "have a next node of last_payday_before_sickness when paid at least 8 weeks of earnings" do
+        add_responses paid_at_least_8_weeks?: "eight_weeks_more"
+        assert_next_node :last_payday_before_sickness?, for_response: "weekly"
+      end
+
+      should "have a next node of last_payday_before_sickness when before payday" do
```
Comment (leenagupte) [inline test/flows/calculate_statutory_sick_pay_test.rb:321]: It's not just for "before_payday", so how about:

```suggestion
      should "have a next node of last_payday_before_sickness when not paid 8 weeks of earnings" do
```

File: test/flows/calculate_statutory_sick_pay_test.rb [hunk 4]
```diff
@@ -0,0 +1,598 @@
...
+    context "validation" do
+      should "not allow sick day to be before 2010" do
+        assert_invalid_response "2009-01-01"
+      end
+
+      should "not allow a sick day in the future" do
+        assert_invalid_response "3000-01-01"
+      end
+
+      should "assert date is more than 8 weeks before last payday before sickness" do
```
Comment (leenagupte) [inline test/flows/calculate_statutory_sick_pay_test.rb:397]:
```suggestion
      should "be invalid if last_payday_before_offset date is more than 8 weeks before last payday before sickness" do
```

File: test/flows/calculate_statutory_sick_pay_test.rb [hunk 5]
```diff
@@ -0,0 +1,598 @@
...
+                    coronavirus_related?: "no",
+                    employee_tell_within_limit?: "yes",
+                    employee_work_different_days?: "no",
+                    first_sick_day?: "2020-01-01",
+                    last_sick_day?: "2020-01-05",
+                    has_linked_sickness?: "yes"
+    end
+
+    context "next_node" do
+      context "setup one" do
```
Comment (leenagupte) [inline test/flows/calculate_statutory_sick_pay_test.rb:538]: What does "setup one" mean? Can it be more descriptive?

**Review by leenagupte — COMMENTED** [review]
Jon, well done for figuring this out! The calculator methods are really difficult to follow so a ⭐  for you!

As mentioned in IRL, there are a few structural things to fix so that this test follows the patterns established in ADR-3 and the other flow tests.  I've mentioned some inline, but here's a summary:

Need to remove the old test files:

test/unit/flows/calculate_statutory_sick_pay_flow_test.rb
test/unit/flows/calculate_statutory_sick_pay_view_test.rb

There's a few format discrepancies as well. I've highlighted a few inline, but the following should apply to the whole file:

1. The context should be named `question: question_name?` if testing a qusetion or `outcome: outcome_name` if testing an outcome.
2. There should be test for "rendering the question" in every context so that we know if the node can be reached.
3. The validation tests should come before the next_node tests.

I'll hold off reviewing the `usual_work_days?` tests until they've been restructured, but on the whole this is looking good!

File: test/flows/calculate_statutory_sick_pay_test.rb [hunk 6]
```diff
@@ -0,0 +1,598 @@
...
+                    employee_work_different_days?: "no"
+    end
+
+    context "next_node" do
+      should "have a next node of last_sick_day for any response" do
+        assert_next_node :last_sick_day?, for_response: "2020-01-01"
+      end
+    end
+
+    context "validation" do
```
[the remark this answers is not in the document]
Comment (JonathanHallam) [reply test/flows/calculate_statutory_sick_pay_test.rb]: Flipped all of these :)

File: test/flows/calculate_statutory_sick_pay_test.rb [hunk 7]
```diff
@@ -0,0 +1,598 @@
...
+        assert_next_node :coronavirus_gp_letter?, for_response: "yes"
+      end
+
+      should "have a next node of employee_tell_within_limit for 'no' response" do
+        assert_next_node :employee_tell_within_limit?, for_response: "no"
+      end
+    end
+  end
+
+  context "coronavirus_gp_letter?" do
```
[the remark this answers is not in the document]
Comment (JonathanHallam) [reply test/flows/calculate_statutory_sick_pay_test.rb]: Added this everywhere so will resolve them all, thanks :)

File: test/flows/calculate_statutory_sick_pay_test.rb (hunk 2 above)
Comment (JonathanHallam) [reply test/flows/calculate_statutory_sick_pay_test.rb]: Added them all 🎉

File: test/flows/calculate_statutory_sick_pay_test.rb (hunk 5 above)
Comment (JonathanHallam) [reply test/flows/calculate_statutory_sick_pay_test.rb]: Sorted

File: test/flows/calculate_statutory_sick_pay_test.rb [hunk 8]
```diff
@@ -0,0 +1,598 @@
...
+    end
+
+    context "next_node" do
+      should "have a next node of last_sick_day for any response" do
+        assert_next_node :last_sick_day?, for_response: "2020-01-01"
+      end
+    end
+
+    context "validation" do
+      should "not allow sick day to be before 2010" do
```
[the remark this answers is not in the document]
Comment (JonathanHallam) [reply test/flows/calculate_statutory_sick_pay_test.rb:154]: This validation comes from a few lines [like this](https://github.com/alphagov/smart-answers/blob/main/app/flows/calculate_statutory_sick_pay_flow.rb#L113) that set the first possible date as the first of January 2010, so I think this is fine

[the remark this answers is not in the document]
Comment (leenagupte) [reply test/flows/calculate_statutory_sick_pay_test.rb:154]: The first possible date is 1st Jan **2011**.

[the remark this answers is not in the document]
Comment (JonathanHallam) [reply test/flows/calculate_statutory_sick_pay_test.rb:154]: ah poop, there's more than one arbitrary date in there! Some are 2010 others are 2011 for reasons! I'll make sure they match

[the remark this answers is not in the document]
Comment (JonathanHallam) [reply test/flows/calculate_statutory_sick_pay_test.rb]: sorted

File: test/flows/calculate_statutory_sick_pay_test.rb (hunk 1 above)
Comment (JonathanHallam) [reply test/flows/calculate_statutory_sick_pay_test.rb]: Sorted all of these 🎉

File: test/flows/calculate_statutory_sick_pay_test.rb (hunk 3 above)
Comment (JonathanHallam) [reply test/flows/calculate_statutory_sick_pay_test.rb]: Good shout, have fixed

File: test/flows/calculate_statutory_sick_pay_test.rb [hunk 9]
```diff
@@ -0,0 +1,726 @@
...
+      context "maximum entitlement reached outcome" do
+        setup do
+          add_responses linked_sickness_start_date?: "2018-12-01",
+                        linked_sickness_end_date?: "2019-12-30",
+                        paid_at_least_8_weeks?: "eight_weeks_less",
+                        total_earnings_before_sick_period?: "70000",
+                        days_covered_by_earnings?: "5"
+        end
+
+        should "have an outcome of not_earned_enough if not enough earned 4" do
```
Comment (leenagupte) [inline test/flows/calculate_statutory_sick_pay_test.rb:647]:
```suggestion
        should "have an outcome of maximum_entitlement_reached for any response if too many previous sick days" do
```

File: test/flows/calculate_statutory_sick_pay_test.rb [hunk 10]
```diff
@@ -0,0 +1,726 @@
...
+          assert_next_node :maximum_entitlement_reached, for_response: "4"
+        end
+      end
+
+      context "entitled to sick pay outcome" do
+        should "have an outcome of not_earned_enough if not enough earned 3" do
```
Comment (leenagupte) [inline test/flows/calculate_statutory_sick_pay_test.rb:653]:
```suggestion
        should "have an outcome of entitled_to_sick_pay for any response if ssp_payment is greater than 0" do
```

File: test/flows/calculate_statutory_sick_pay_test.rb [hunk 11]
```diff
@@ -0,0 +1,726 @@
...
+      context "maximum entitlement reached v2 outcome" do
+        setup do
+          add_responses linked_sickness_start_date?: "2019-05-23",
+                        linked_sickness_end_date?: "2019-12-29",
+                        paid_at_least_8_weeks?: "eight_weeks_less",
+                        total_earnings_before_sick_period?: "70000",
+                        days_covered_by_earnings?: "1"
+        end
+
+        should "have an outcome of not_earned_enough if not enough earned 4" do
```
Comment (leenagupte) [inline test/flows/calculate_statutory_sick_pay_test.rb:667]:
```suggestion
        should "have an outcome of maximum_entitlement_reached for any response if zero days can be paid for period" do
```

File: test/flows/calculate_statutory_sick_pay_test.rb [hunk 12]
```diff
@@ -0,0 +1,726 @@
...
+          assert_next_node :maximum_entitlement_reached, for_response: "0"
+        end
+      end
+
+      context "not entitled 3 days not paid outcome" do
+        should "have an outcome of not_earned_enough if not enough earned 2" do
```
Comment (leenagupte) [inline test/flows/calculate_statutory_sick_pay_test.rb:673]:
```suggestion
        should "have an outcome of not_entitled_3_days_not_paid for any response if sick for less than three days" do
```

**Review by leenagupte — COMMENTED** [review]
I think the tests are looking pretty good. There's just a bit of clean up required. For example, there are some missing `?` from the test names and contexts, and some of the test names could be re-worded to be more clear about what is being tested (some are copy and paste errors).

File: test/flows/calculate_statutory_sick_pay_test.rb [hunk 13]
```diff
@@ -0,0 +1,702 @@
...
+    end
+
+    should "render the question" do
+      assert_rendered_question
+    end
+
+    context "next_node" do
+      should "have an outcome of not_earned_enough for any response if employee_average_weekly_earnings is less than lower_earning_limit_rate" do
+        add_responses linked_sickness_start_date?: "2019-12-01",
+                      linked_sickness_end_date?: "2019-12-30",
```
Comment (leenagupte) [inline test/flows/calculate_statutory_sick_pay_test.rb:626]: Minor: don't need to set these two values as they are the same as the default. The same applies to `paid_at_least_8_weeks?` and `days_covered_by_earnings?`

The only value that needs to be updated is: `total_earnings_before_sick_period?`

File: test/flows/calculate_statutory_sick_pay_test.rb [hunk 14]
```diff
@@ -0,0 +1,702 @@
...
+
+        assert_next_node :not_earned_enough, for_response: "4"
+      end
+
+      should "have an outcome of maximum_entitlement_reached for any response if too many previous sick days" do
+        add_responses linked_sickness_start_date?: "2018-12-01",
+                      linked_sickness_end_date?: "2019-12-30",
+                      paid_at_least_8_weeks?: "eight_weeks_less",
+                      total_earnings_before_sick_period?: "70000",
+                      days_covered_by_earnings?: "5"
```
Comment (leenagupte) [inline test/flows/calculate_statutory_sick_pay_test.rb:639]: Same here. I don't think these values need to be set again as they are the same as the defaults in the `setup`.

File: test/flows/calculate_statutory_sick_pay_test.rb [hunk 15]
```diff
@@ -0,0 +1,702 @@
...
+
+      should "have an outcome of entitled_to_sick_pay for any response if ssp_payment is greater than 0" do
+        assert_next_node :entitled_to_sick_pay, for_response: "3"
+      end
+
+      should "have an outcome of maximum_entitlement_reached for any response if zero days can be paid for period" do
+        add_responses linked_sickness_start_date?: "2019-05-23",
+                      linked_sickness_end_date?: "2019-12-29",
+                      paid_at_least_8_weeks?: "eight_weeks_less",
+                      total_earnings_before_sick_period?: "70000",
```
Comment (leenagupte) [inline test/flows/calculate_statutory_sick_pay_test.rb:652]: Same here. I don't think these values need to be set again as they are the same as the defaults in the setup.

**Review by leenagupte — COMMENTED** [review]
Thanks for squishing all of the contexts @Rosa-Fox ⭐ 

@JonathanHallam there are a few responses being set unnecessarily in the tests. Removing would drop the test file under 700 lines 😱. Once they're removed this should be good to go.

File: test/flows/calculate_statutory_sick_pay_test.rb (hunk 13 above)
Comment (JonathanHallam) [reply test/flows/calculate_statutory_sick_pay_test.rb]: Have sorted all of these and minimised as much as possible :)

**Comment by JonathanHallam** [comment]
@leenagupte this is ready for re-review now! 🤞 everything's good :)

**Review by leenagupte — APPROVED** [review]
🏅 LGTM! Thanks for making all of the updates.

**Pull request merged**